org.eclipse.m2e.core/src/org/eclipse/m2e/core/project/configurator/AbstractProjectConfigurator.java

```diff
@@ -167,15 +175,43 @@ public static void addNature(IProject project, String natureId, int updateFlags,
       throws CoreException {
     if(!project.hasNature(natureId)) {
       IProjectDescription description = project.getDescription();
-      String[] prevNatures = description.getNatureIds();
-      String[] newNatures = new String[prevNatures.length + 1];
-      System.arraycopy(prevNatures, 0, newNatures, 1, prevNatures.length);
-      newNatures[0] = natureId;
-      description.setNatureIds(newNatures);
+      var natures = Stream.concat(Stream.of(natureId), Arrays.stream(description.getNatureIds()));
+      description.setNatureIds(natures.toArray(String[]::new));
       project.setDescription(description, updateFlags, monitor);
     }
   }
 
+  /**
+   * Creates a problem marker in the pom.xml file at the specified element of the given Plugin-execution with the passed
+   * message and severity.
+   * <p>
+   * If the attribute of the execution is specified in a parent pom.xml outside of the workspace the marker is created
+   * at the parent-element of the request's project.
+   * </p>
+   * 
+   * @param execution the execution
+   * @param element the XML-element to mark
+   * @param request the request of the project being build
+   * @param problemSeverity the problems severity, one of {@link IMarker#SEVERITY_INFO SEVERITY_INFO},
+   *          {@link IMarker#SEVERITY_WARNING SEVERITY_WARNING} or {@link IMarker#SEVERITY_ERROR SEVERITY_ERROR}
+   * @param problemMessage the message of the created problem marker
+   */
+  protected void createProblemMarker(MojoExecution execution, String element, ProjectConfigurationRequest request,
```

Comment:
Could the IMarker be returned here?
What if one wants to create the marker on the value of the element? Can't we give it a text position instead, and an helper to locate an element in a mojoExecution?

Reply:
> Could the IMarker be returned here? What if one wants to create the marker on the value of the element? 

In general `IMavenMarkerManager` could be changed to return the marker too.
Creating the marker at the value is not yet supported, but in general could be considered.

> Can't we give it a text position instead, and an helper to locate an element in a mojoExecution?

Yes Christoph suggested that as well before.
And after some rethinking about it, it is probably the better way for a generalization because IIRC for markers of custom Ids m2e just creates them using the Eclipse Resources API. So there is no benefit in providing an extra API for that. But I have to check that again.
However I think for a simple case a convenient API method like this would be useful nevertheless.

Reply:
> However I think for a simple case a convenient API method like this would be useful nevertheless.

There is uncertainity with this API that needs to be covered before it can be claimed "API-quality" (ie we won't want to change it for a long time), and this comes by nature of using the "element" name:
* How to place a marker on the value or a subset of the value? Or on some attribute (some mojo configuration may like it?
* What happens is the element is repeated in the configuration?
* What happens if there is no such element in the configuration?
* ...

All those questions wouldn't be raised if we clearly separate locating the marker vs creating it. If it is clear that locating the marker is the important part, and not creating it (because Platform API is enough), then instead of API to create a marker, we should have an API to reoslve locations and use already existing satisfying APIs from Platform to create it.